src/github/command.rs

```diff
@@ -0,0 +1,445 @@
+
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub enum Label {
+    FFCP,
+    PFCP,
+    FCP,
+    Postponed,
+    Closed,
+    DispositionMerge,
+    DispositionClose,
+    DispositionPostpone,
+}
+
+impl Label {
+    pub fn as_str(self) -> &'static str {
+        use self::Label::*;
+        match self {
+            FFCP => "finished-final-comment-period",
+            PFCP => "proposed-final-comment-period",
+            FCP => "final-comment-period",
+            Postponed => "postponed",
+            Closed => "closed",
+            DispositionMerge => "disposition-merge",
+            DispositionClose => "disposition-close",
+            DispositionPostpone => "disposition-postpone",
+        }
+    }
+}
+
+impl fmt::Display for Label {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.write_str(self.as_str())
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum RfcBotCommand<'a> {
+    FcpPropose(FcpDisposition),
+    FcpCancel,
+    Reviewed,
+    NewConcern(&'a str),
+    ResolveConcern(&'a str),
+    FeedbackRequest(&'a str),
+    AskQuestion {
+        teams: BTreeSet<&'a str>,
+        question: &'a str,
+    },
+}
+
+impl<'a> RfcBotCommand<'a> {
+    pub fn from_str_all(command: &'a str) -> impl Iterator<Item = RfcBotCommand<'a>> {
+        // Get the tokens for each command line (starts with a bot mention)
+        command.lines()
+               .filter(|&l| l.starts_with(RFC_BOT_MENTION))
+               .map(from_invocation_line)
+               .filter_map(Result::ok)
+    }
+
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum FcpDisposition {
+    Merge,
+    Close,
+    Postpone,
+}
+
+const FCP_REPR_MERGE: &'static str = "merge";
+const FCP_REPR_CLOSE: &'static str = "close";
+const FCP_REPR_POSTPONE: &'static str = "postpone";
+
+impl FcpDisposition {
+    pub fn repr(self) -> &'static str {
+        match self {
+            FcpDisposition::Merge => FCP_REPR_MERGE,
+            FcpDisposition::Close => FCP_REPR_CLOSE,
+            FcpDisposition::Postpone => FCP_REPR_POSTPONE,
+        }
+    }
+
+    pub fn from_str(string: &str) -> DashResult<Self> {
+        Ok(match string {
+            FCP_REPR_MERGE => FcpDisposition::Merge,
+            FCP_REPR_CLOSE => FcpDisposition::Close,
+            FCP_REPR_POSTPONE => FcpDisposition::Postpone,
+            _ => throw!(DashError::Misc(None)),
+        })
+    }
+
+    pub fn label(self) -> Label {
+        match self {
+            FcpDisposition::Merge => Label::DispositionMerge,
+            FcpDisposition::Close => Label::DispositionClose,
+            FcpDisposition::Postpone => Label::DispositionPostpone,
+        }
+    }
+}
+
+/// Parses the text of a subcommand.
+fn parse_command_text<'a>(command: &'a str, subcommand: &'a str) -> &'a str {
+    let name_start = command.find(subcommand).unwrap() + subcommand.len();
+    command[name_start..].trim()
+}
+
+fn strip_prefix<'h>(haystack: &'h str, prefix: &str) -> &'h str {
```

Comment:
Is this significantly different from `trim_left_matches`?

Reply:
> Returns a string slice with all prefixes that match a pattern repeatedly removed.

Yep, this removes one occurrence, `trim_left_matches` removes an arbitrary number :)
I really wanted to use `trim_left_matches` but was sorry to see it wasn't what I wanted.

Reply:
Maybe [str::splitn().1](https://doc.rust-lang.org/std/primitive.str.html#method.splitn) would work?

Reply:
Went with https://doc.rust-lang.org/std/primitive.str.html#method.split_terminator instead :)

Reply:
I spoke too soon... that didn't work.

With `splitn` it becomes:

```rust
fn strip_prefix_2<'h>(haystack: &'h str, prefix: &str) -> &'h str {
    let mut iter = haystack.splitn(2, prefix);
    let _ = iter.next();
    iter.next().unwrap_or(haystack).trim()
}
```

which is less clear to me than the current implementation so I'll stick with the code as-is. :)